File: src/jp/togashi/android/kkmush/KKMushActivity.java
package jp.togashi.android.kkmush;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.text.ClipboardManager;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.SeekBar;
import android.widget.TextView;
import android.widget.SeekBar.OnSeekBarChangeListener;

public class KKMushActivity extends Activity implements OnSeekBarChangeListener {
    private static final String ACTION_INTERCEPT = "com.adamrocker.android.simeji.ACTION_INTERCEPT";
    private static final String REPLACE_KEY = "replace_key";
    
    private static final String PREF_KEY_LEVEL = "pref_key_level";
    
    private static final Pattern mGobi = Pattern.compile("([^a-zA-Z0-9 　。ッ]+)([ 　。!！\\?？])");
    private static final Pattern mBunmatsu = Pattern.compile(".*[^a-z0-9 　。ッ]$");
    
    private String mInputStr = "";
    private String mOutputStr;
    
    private String convert(final String src, final int level) {
        String outStr = new String(src);
        
        if (level >= 1) {
            outStr = outStr.replace("ん", "ン");
        }
        
        if (level >= 2) {
            outStr = outStr.replace("っ", "ッ");
        }
        
        if (level >= 3) {
            Matcher m = mGobi.matcher(outStr);
            outStr = m.replaceAll("$1ッ$2");
        }
        
        if (level >= 4) {
            Matcher m = mBunmatsu.matcher(outStr);
            if (m.matches()) {
                outStr += "ッ";
            }
        }
        
        return outStr;
    }
    
    private int getLevel() {
        int lv = 1;
        SeekBar sb = (SeekBar)findViewById(R.id.seekBar1);
        if (sb != null) {
            lv = sb.getProgress() + 1;
        }
        return lv;
    }
    
    private void update() {
        int lv = getLevel();
        
        TextView tv = (TextView)findViewById(R.id.beforetext);
        if (tv != null) {
            tv.setText(mInputStr);
        }
        
        tv = (TextView)findViewById(R.id.koikelevel);
        if (tv != null) {
            tv.setText(String.format(getResources().getString(R.string.koikelevel), lv));
        }
        
        String after = convert(mInputStr, lv);
        
        tv = (TextView)findViewById(R.id.aftertext);
        if (tv != null) {
            tv.setText(after);
        }
    }
    
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        
        Intent input = getIntent();
        String action = input.getAction();
        if (action != null && ACTION_INTERCEPT.equals(action)) {
            mInputStr = input.getStringExtra(REPLACE_KEY);
        }
        
        if (TextUtils.isEmpty(mInputStr)) {
            ClipboardManager cbm = (ClipboardManager)getSystemService(CLIPBOARD_SERVICE);
            CharSequence cs = cbm.getText();
            if (cs != null) {
                mInputStr = cs.toString();
            }
        }
        
        if (TextUtils.isEmpty(mInputStr)) {
            finish();
            return;
        }
        mOutputStr = new String(mInputStr);
        
        setContentView(R.layout.main);
        
        SeekBar sb = (SeekBar)findViewById(R.id.seekBar1);
        if (sb != null) {
            sb.setOnSeekBarChangeListener(this);
        }
        
        Button btn = (Button)findViewById(R.id.okbutton);
        if (btn != null) {
            btn.setOnClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    mOutputStr = convert(mInputStr, getLevel());
                    finish();
                }
            });
        }
    }
    
    @Override
    public void finish() {
        if (!TextUtils.isEmpty(mOutputStr)) {
            Intent output = new Intent();
            output.putExtra(REPLACE_KEY, mOutputStr);
            setResult(RESULT_OK, output);
            ClipboardManager cbm = (ClipboardManager)getSystemService(CLIPBOARD_SERVICE);
            cbm.setText(mOutputStr);
        }
        super.finish();
    }
    
    @Override
    public void onResume() {
        super.onResume();
        SeekBar sb = (SeekBar)findViewById(R.id.seekBar1);
        if (sb != null) {
            SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
            int savedLevel = prefs.getInt(PREF_KEY_LEVEL, 1);
            sb.setProgress(savedLevel);
        }
        update();
    }
    
    @Override
    public void onPause() {
        super.onPause();
        SeekBar sb = (SeekBar)findViewById(R.id.seekBar1);
        if (sb != null) {
            SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
            int savedLevel = prefs.getInt(PREF_KEY_LEVEL, 1);
            int currentLevel = sb.getProgress();
            if (savedLevel != currentLevel) {
                Editor editor = prefs.edit();
                editor.putInt(PREF_KEY_LEVEL, currentLevel);
                editor.commit();
            }
        }
    }
    
    @Override
    public void onProgressChanged(SeekBar seekBar, int progress,
            boolean fromUser) {
        update();
    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {
    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {
    }
}

File: app/src/main/java/jp/togashi/android/kkmush/KKMushActivity.kt
package jp.togashi.android.kkmush

import android.app.Activity
import android.app.Dialog
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDialogFragment
import androidx.preference.PreferenceManager
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.ktx.analytics
import com.google.firebase.analytics.ktx.logEvent
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.google.firebase.ktx.Firebase
import jp.togashi.android.kkmush.databinding.MainBinding
import timber.log.Timber

class KKMushActivity : AppCompatActivity() {

    companion object {
        private const val ACTION_INTERCEPT = "com.adamrocker.android.simeji.ACTION_INTERCEPT"
        private const val REPLACE_KEY = "replace_key"

        private const val PREF_KEY_LEVEL = "pref_key_level"
    }

    init {
        Timber.plant(
                if (BuildConfig.DEBUG) Timber.DebugTree()
                else object: Timber.Tree() {
                    override fun log(priority: Int, tag: String?, message: String, t: Throwable?) {
                        val priorityLabel = when (priority) {
                            Log.VERBOSE, Log.DEBUG -> return
                            Log.INFO -> "I"
                            Log.WARN -> "W"
                            Log.ERROR -> "E"
                            else -> "I"
                        }
                        FirebaseCrashlytics.getInstance().apply {
                            log("$priorityLabel/${tag ?: "NONE"}: $message")
                            t?.let {
                                recordException(it)
                            }
                        }
                    }
                }
        )
    }

    private lateinit var firebaseAnalytics: FirebaseAnalytics

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (!hasFocus) return

        if (viewModel.source.isNullOrEmpty()) {
            val cbm = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            viewModel.source = cbm.primaryClip?.getItemAt(0)?.text.toString()
        }
    }

    private lateinit var binding: MainBinding
    private lateinit var viewModel: KKMushViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        firebaseAnalytics = Firebase.analytics

        viewModel = KKMushViewModel(this)

        if (intent.action == ACTION_INTERCEPT) {
            viewModel.source = intent.getStringExtra(REPLACE_KEY)
        }

        if (viewModel.source.isNullOrEmpty()) {
            val cbm = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            viewModel.source = cbm.primaryClip?.getItemAt(0)?.text?.toString()
        }

        setContentView(MainBinding.inflate(LayoutInflater.from(this)).also {
            it.viewModel = viewModel
            it.lifecycleOwner = this
            binding = it
        }.root)

        binding.buttonOk.setOnClickListener {
            viewModel.lastResult?.let {
                firebaseAnalytics.logEvent("convert") {
                    it.counts.keys.forEach { k ->
                        param(k, it.counts[k]?.toLong() ?: 0)
                    }
                }
            }
            finish()
        }
    }

    override fun finish() {
        if (!TextUtils.isEmpty(viewModel.converted)) {
            val output = Intent()
            output.putExtra(REPLACE_KEY, viewModel.converted)
            setResult(Activity.RESULT_OK, output)
            val cbm = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            cbm.setPrimaryClip(ClipData.newPlainText(getString(R.string.after_text), viewModel.converted))
        }
        super.finish()
    }

    public override fun onResume() {
        super.onResume()
        val prefs = PreferenceManager.getDefaultSharedPreferences(applicationContext)
        val savedLevel = prefs.getInt(PREF_KEY_LEVEL, 1)
        viewModel.koikeLevel = savedLevel
    }

    public override fun onPause() {
        super.onPause()
        PreferenceManager.getDefaultSharedPreferences(applicationContext).apply {
            edit().putInt(PREF_KEY_LEVEL, viewModel.koikeLevel).apply()
        }
    }

//    private val seekBarChangeListener = object: OnSeekBarChangeListener {
//        override fun onProgressChanged(p0: SeekBar?, p1: Int, p2: Boolean) {
//            update()
//        }
//
//        override fun onStartTrackingTouch(p0: SeekBar?) {
//        }
//
//        override fun onStopTrackingTouch(p0: SeekBar?) {
//        }
//    }
//
    class AboutDialogFragment: AppCompatDialogFragment() {
        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
            return AlertDialog.Builder(requireContext()).apply {
                setTitle(R.string.menu_about)
                setIcon(R.mipmap.ic_launcher_round)
                setView(View.inflate(context, R.layout.about, null).apply {
                    (findViewById<TextView>(R.id.versionTextView))?.text = BuildConfig.VERSION_NAME
                })
                setPositiveButton(R.string.button_dismiss) { dialog, _ -> dialog.dismiss() }
            }.create()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when(item.itemId) {
            R.id.menu_about -> {
                AboutDialogFragment().show(supportFragmentManager, "TAG_DIALOG")
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

}
